Index: engine/apps/engine/src/java/com/occamlab/te/TECore.java
===================================================================
--- engine/apps/engine/src/java/com/occamlab/te/TECore.java	(revisión: 259)
+++ engine/apps/engine/src/java/com/occamlab/te/TECore.java	(copia de trabajo)
@@ -501,7 +501,11 @@
             }
 
             // Set headers
-	    String mid = ((Element) body).getAttribute("mid");
+            String mid = null;
+            if ( body != null ) {
+                mid = ((Element) body).getAttribute("mid");    
+            }
+            
             if (mid != "" && mid != null) {
                 uc.setRequestProperty("Message-ID", mid);
             }
